File: mvn_java_proxy/src/main/java/mystdeim/vertx_examples/mvn_java_proxy/service/AccountService.java
package mystdeim.vertx_examples.mvn_java_proxy.service;

import io.vertx.codegen.annotations.GenIgnore;
import io.vertx.codegen.annotations.ProxyGen;
import io.vertx.codegen.annotations.VertxGen;
import io.vertx.core.AsyncResult;
import io.vertx.core.Future;
import io.vertx.core.Handler;
import io.vertx.core.Vertx;
import mystdeim.vertx_examples.mvn_java_proxy.model.Account;
import mystdeim.vertx_examples.mvn_java_proxy.service.impl.AccountServiceImpl;

/**
 * @author Roman Novikov
 */
@ProxyGen
@VertxGen
public interface AccountService {

    void create(Account account, Handler<AsyncResult<Long>> handler);
    void get(long id, Handler<AsyncResult<Account>> handler);

    String ADDRESS = "account";

    @GenIgnore
    default Future<Account> get(long id) {
        Future<Account> f = Future.future();
        get(id, f.completer());
        return f;
    }

    @GenIgnore
    default Future<Long> create(Account account) {
        Future<Long> f = Future.future();
        create(account, f.completer());
        return f;
    }

    static AccountService createService(Vertx vertx) {
        return new AccountServiceImpl(vertx);
    }

    static AccountService createProxy(Vertx vertx, String address) {
        return new AccountServiceVertxEBProxy(vertx, address);
    }
}

File: mvn_kotlin_proxy/src/main/kotlin/mystdeim/vertx_examples/mvn_kotlin_proxy/service/AccountService.kt
package mystdeim.vertx_examples.mvn_kotlin_proxy.service

import io.vertx.codegen.annotations.GenIgnore
import io.vertx.codegen.annotations.ProxyGen
import io.vertx.codegen.annotations.VertxGen
import io.vertx.core.AsyncResult
import io.vertx.core.Handler
import io.vertx.core.Vertx
import mystdeim.vertx_examples.mvn_kotlin_proxy.model.Account
import mystdeim.vertx_examples.mvn_kotlin_proxy.service.impl.AccountServiceImpl

/**
 * @author Roman Novikov
 */
@ProxyGen
@VertxGen
interface AccountService {

    fun create(account: Account, handler: Handler<AsyncResult<Long>>)
    operator fun get(id: Long, handler: Handler<AsyncResult<Account>>)

    @GenIgnore
    companion object {

        val ADDRESS = "account"

        fun createService(vertx: Vertx): AccountService {
            return AccountServiceImpl(vertx)
        }

        fun createProxy(vertx: Vertx, address: String): AccountService {
            return AccountServiceVertxEBProxy(vertx, address)
        }
    }
}